File: app/src/main/java/data/room/BaseDao.java
package data.room;

import android.arch.persistence.room.*;

public interface BaseDao<T> {
    /**
     * Inserts the object into the database.
     * @param obj the object to insert into the database
     */
    @Insert
    void insert(T obj);

    /**
     * Updates the object in the database, if it exists. Replaces on conflict.
     * @param obj the updated object
     */
    @Update(onConflict = OnConflictStrategy.REPLACE)
    void update(T obj);

    /**
     * Deletes the object in the database, if it exists.
     * @param obj the object to find and delete from the database
     */
    @Delete
    void delete(T obj);

}

File: app/src/main/java/data/room/dao/BaseDao.kt
package data.room.dao

import androidx.room.*

interface BaseDao<T> {
    /**
     * Inserts the object into the database. Replaces on conflict
     * @param obj the object to insert into the database
     */
    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insert(obj: T): Long

    /**
     * Updates the object in the database, if it exists. Replaces on conflict.
     * @param obj the updated object
     */
    @Update(onConflict = OnConflictStrategy.IGNORE)
    suspend fun update(obj: T)

    /**
     * Deletes the object in the database, if it exists.
     * @param obj the object to find and delete from the database
     */
    @Delete
    suspend fun delete(obj: T)
}
